xls/modules/rle/rle.x

```diff
@@ -0,0 +1,306 @@
+    symbol_r: chan<bits[SYMBOL_WIDTH]> in;
+    // input with flush request (should be read together with symbol_r)
+    flush_r: chan<bool> in;
+    // output with data, contains symbol and symbol count
+    data_s: chan<EncOutData<SYMBOL_WIDTH, COUNT_WIDTH>> out;
+
+    init {(
+        RLEEncState<SYMBOL_WIDTH, COUNT_WIDTH> {
+            prev: EncOutData<SYMBOL_WIDTH, COUNT_WIDTH> {
+                symbol: bits[SYMBOL_WIDTH]:0,
+                count: bits[COUNT_WIDTH]:0,
+            },
+            flush: false,
+        }
+    )}
+
+    config (
+        symbol_r: chan<bits[SYMBOL_WIDTH]> in,
+        flush_r: chan<bool> in,
+        data_s: chan<EncOutData<SYMBOL_WIDTH, COUNT_WIDTH>> out,
+    ) {(symbol_r, flush_r, data_s)}
+
+    next (tok: token, state: RLEEncState<SYMBOL_WIDTH, COUNT_WIDTH>) {
+        let (flush_tok, flush) = recv_if(tok, flush_r, !state.flush, false);
+        let (symbol_tok, symbol) = recv_if(tok, symbol_r, !state.flush, bits[SYMBOL_WIDTH]:0);
+        let recv_tok = join(flush_tok, symbol_tok);
+
+        let prev_symbol_valid = state.prev.count != bits[COUNT_WIDTH]:0;
+        let symbol_differ = prev_symbol_valid && (symbol != state.prev.symbol);
+        let overflow = state.prev.count == std::unsigned_max_value<COUNT_WIDTH>();
+
+        let next_state = if (state.flush) {
+            RLEEncState {
+                prev: EncOutData<SYMBOL_WIDTH, COUNT_WIDTH> {
+                    symbol: bits[SYMBOL_WIDTH]:0,
+                    count: bits[COUNT_WIDTH]:0,
+                },
+                flush: flush,
+            }
+        } else if (symbol_differ || overflow) {
+            RLEEncState {
+                prev: EncOutData<SYMBOL_WIDTH, COUNT_WIDTH> {
+                    symbol: symbol,
+                    count: bits[COUNT_WIDTH]:1,
+                },
+                flush: flush,
+            }
+        } else {
+            RLEEncState {
+                prev: EncOutData<SYMBOL_WIDTH, COUNT_WIDTH> {
+                    symbol: symbol,
+                    count: state.prev.count + bits[COUNT_WIDTH]:1,
+                },
+                flush: flush,
+            }
+        };
+
+        let do_send = state.flush || symbol_differ || overflow;
+        let data_tok = send_if(recv_tok, data_s, do_send, state.prev);
+        next_state
+    }
+}
+
+// RLE encoder specialization for the codegen
+proc RLEEnc32 {
+
+    init {()}
+
+    config (
+        symbol_r: chan<bits[32]> in,
+        flush_r: chan<bool> in,
+        data_s: chan<EncOutData<u32:32, u32:2>> out,
+    ) {
+        spawn RLEEnc<u32:32, u32:2>(symbol_r, flush_r, data_s);
+        ()
+    }
+
+    next (tok: token, state: ()) {
+        ()
+    }
+}
+
+// Tests
+
+type SendDataPair = (bits[32], bool);     // (symbol, flush)
+type RecvDataPair = (bits[32], bits[2]);  // (symbol, count)
+
+const TEST_SEND_LEN = u32:20;
+const TEST_RECV_LEN = u32:10;
+
+const TEST_SEND_LOOKUP = SendDataPair[TEST_SEND_LEN]:[
```

Comment:
if you want to do table-based testing, you might consider interleaving input and output in a `struct` so that it can read more fluently.

Reply:
The problem lies in the fact that not every send request has a corresponding receive request. So in general there will be empty spaces in the struct that collect the transactions, plus some sort of flag indicating that the receive is expected would be required. 

At the same time, there is a need to check for the send that has a flush flag set. I'm not sure whether having one struct for everything could simplify the current mechanism without providing much more complexity in the structure of data. 

But maybe I don't see something obvious here. Do you have anything specific in mind?

Reply:
Having the `last` signal exposed as an output as described in https://github.com/google/xls/pull/974#discussion_r1209390982, simplifies also the checks of the data received from the encoder.

Reply:
Is the current solution acceptable for you @proppy?